From 14e88a4858f2a81b7e2743fde4a9789bbaaecf28 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Mon, 27 Apr 2020 12:24:52 +0800
Subject: [PATCH 56/62] MLK-23694-10 drm/imx: sec-dsim: NULL panel ptr when
 host detached

The panel ptr variable should be NULL in the DSI host's
detach() hook instead of in the unbind() to fix below
warning:

[    2.322570] ------------[ cut here ]------------
[    2.327193] WARNING: CPU: 0 PID: 188 at drivers/gpu/drm/bridge/sec-dsim.c:560 sec_mipi_dsim_host_detach+0x28/0x38
[    2.337452] Modules linked in:
[    2.340510] CPU: 0 PID: 188 Comm: kworker/0:2 Not tainted 5.4.24-04741-g8372cb5b3eec-dirty #1524
[    2.349293] Hardware name: FSL i.MX8MP EVK (DT)
[    2.353828] Workqueue: events deferred_probe_work_func
[    2.358967] pstate: 00000005 (nzcv daif -PAN -UAO)
[    2.363759] pc : sec_mipi_dsim_host_detach+0x28/0x38
[    2.368728] lr : mipi_dsi_detach+0x24/0x38
[    2.372823] sp : ffff80001232b5a0
[    2.376135] x29: ffff80001232b5a0 x28: ffff8000119b6000
[    2.381446] x27: 0000000000000000 x26: 00000000fffffdfb
[    2.386758] x25: 0000000000000001 x24: ffff8000114f96b8
[    2.392071] x23: ffff0001763fd810 x22: ffff0001769eb800
[    2.397384] x21: ffff800011a256e8 x20: ffff000177909780
[    2.402696] x19: ffff00017793a000 x18: 0000000000000010
[    2.408008] x17: ffff800011ded000 x16: 0000000000000000
[    2.413320] x15: 0000000000000004 x14: 73706f5f786d695f
[    2.418632] x13: 0000000000000000 x12: ffff00017457dc28
[    2.423945] x11: ffff00017457db60 x10: 0000000000000000
[    2.429257] x9 : ffff000176020e98 x8 : ffff000176020e90
[    2.434570] x7 : 00000000400c0000 x6 : 0000000000210d00
[    2.439882] x5 : 0000000000000001 x4 : 0000000000000000
[    2.445194] x3 : ffff00017793a0dc x2 : ffff80001073b5b8
[    2.450504] x1 : ffff0001778e3080 x0 : 0000000000000000
[    2.455818] Call trace:
[    2.458266]  sec_mipi_dsim_host_detach+0x28/0x38
[    2.462885]  rad_panel_remove+0x18/0x68
[    2.466720]  mipi_dsi_drv_remove+0x1c/0x28
[    2.470818]  device_release_driver_internal+0xf4/0x1c0
[    2.475955]  device_release_driver+0x14/0x20
[    2.480226]  bus_remove_device+0x118/0x128
[    2.484323]  device_del+0x13c/0x348
[    2.487810]  device_unregister+0x14/0x30
[    2.491736]  mipi_dsi_remove_device_fn+0xc/0x18
[    2.496270]  device_for_each_child+0x58/0xa0
[    2.500539]  mipi_dsi_host_unregister+0x38/0x70
[    2.505069]  sec_mipi_dsim_unbind+0x30/0x40
[    2.509254]  imx_sec_dsim_unbind+0x24/0x48
[    2.513351]  component_unbind.isra.0+0x2c/0x50
[    2.517793]  component_bind_all+0x1e0/0x228
[    2.521976]  imx_drm_bind+0xb8/0x150
[    2.525551]  try_to_bring_up_master+0x164/0x1c0
[    2.530082]  __component_add+0xa0/0x168
[    2.533917]  component_add+0x10/0x18
[    2.537493]  lcdifv3_crtc_probe+0x4c/0x78
[    2.541504]  platform_drv_probe+0x50/0xa0
[    2.545514]  really_probe+0xd4/0x318
[    2.549088]  driver_probe_device+0x54/0xe8
[    2.553183]  __device_attach_driver+0x80/0xb8
[    2.557542]  bus_for_each_drv+0x74/0xc0
[    2.561376]  __device_attach+0xdc/0x138
[    2.565213]  device_initial_probe+0x10/0x18
[    2.569396]  bus_probe_device+0x90/0x98
[    2.573233]  device_add+0x378/0x648
[    2.576721]  platform_device_add+0xfc/0x228
[    2.580908]  imx_lcdifv3_probe+0x2b0/0x388
[    2.585003]  platform_drv_probe+0x50/0xa0
[    2.589013]  really_probe+0xd4/0x318
[    2.592592]  driver_probe_device+0x54/0xe8
[    2.596689]  __device_attach_driver+0x80/0xb8
[    2.601045]  bus_for_each_drv+0x74/0xc0
[    2.604882]  __device_attach+0xdc/0x138
[    2.608721]  device_initial_probe+0x10/0x18
[    2.612906]  bus_probe_device+0x90/0x98
[    2.616740]  deferred_probe_work_func+0x64/0x98
[    2.621274]  process_one_work+0x198/0x320
[    2.625284]  worker_thread+0x1f0/0x420
[    2.629036]  kthread+0xf0/0x120
[    2.632179]  ret_from_fork+0x10/0x18
[    2.635753] ---[ end trace f85c345edc4c5da8 ]---

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Reviewed-by: Liu Ying <victor.liu@nxp.com>
---
 drivers/gpu/drm/bridge/sec-dsim.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/bridge/sec-dsim.c b/drivers/gpu/drm/bridge/sec-dsim.c
index 8af5b7787999..c2247afc506d 100644
--- a/drivers/gpu/drm/bridge/sec-dsim.c
+++ b/drivers/gpu/drm/bridge/sec-dsim.c
@@ -565,6 +565,9 @@ static int sec_mipi_dsim_host_detach(struct mipi_dsi_host *host,
 	dsim->format	 = 0;
 	dsim->mode_flags = 0;
 
+	/* detached panel should be NULL */
+	dsim->panel = NULL;
+
 	return 0;
 }
 
@@ -1965,7 +1968,6 @@ void sec_mipi_dsim_unbind(struct device *dev, struct device *master, void *data)
 	if (dsim->panel) {
 		drm_panel_detach(dsim->panel);
 		drm_connector_cleanup(&dsim->connector);
-		dsim->panel = NULL;
 	}
 
 	mipi_dsi_host_unregister(&dsim->dsi_host);
-- 
2.17.1

